lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,341 @@
+        [&stringRangeValue](const auto &mem) {
+          return ((!std::get<0>(mem) && !std::get<0>(stringRangeValue)) ||
+              (!std::get<1>(mem) && !std::get<1>(stringRangeValue)) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<0>(stringRangeValue) &&
+                       ((std::get<0>(stringRangeValue).value())
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<0>(stringRangeValue) &&
+                          ((std::get<0>(stringRangeValue).value())
+                                  .compare(std::get<1>(mem).value()) <= 0)))) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<1>(stringRangeValue) &&
+                       (std::get<1>(stringRangeValue)
+                               .value()
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<1>(stringRangeValue) &&
+                          (std::get<1>(stringRangeValue)
+                                  .value()
+                                  .compare(std::get<1>(mem).value()) <= 0)))));
+        });
+
+    if (selectCaseStringRangeValsIter == selectCaseStringRangeVals.end()) {
+      selectCaseStringRangeVals.push_back(stringRangeValue);
+    } else {
+      ReportOverlapCaseValError(
+          std::get<2>(*selectCaseStringRangeValsIter), source);
+    }
+  }
+
+private:
+  using selectCaseStmtTypes =
+      std::vector<std::pair<std::variant<std::int64_t, bool, std::string>,
+          parser::CharBlock>>;
+  selectCaseStmtTypes selectCaseVals;
+  selectCaseStmtTypes::iterator selectCaseValsIter;
+  using selectCaseIntRangeType =
+      std::vector<std::tuple<std::optional<std::int64_t>,
+          std::optional<std::int64_t>, parser::CharBlock>>;
+  selectCaseIntRangeType selectCaseIntRangeVals;
+  selectCaseIntRangeType::iterator selectCaseIntRangeValsIter;
+  using selectCaseStringRangeType =
+      std::vector<std::tuple<std::optional<std::string>,
+          std::optional<std::string>, parser::CharBlock>>;
+  selectCaseStringRangeType selectCaseStringRangeVals;
+  selectCaseStringRangeType::iterator selectCaseStringRangeValsIter;
+  parser::Messages &messages_;
+};
+
+void SelectStmtChecker::Leave(
+    const parser::CaseConstruct &selectCaseConstruct) {
+  const auto &selectCaseStmt{
+      std::get<parser::Statement<parser::SelectCaseStmt>>(
+          selectCaseConstruct.t)};
+  const auto &parsedExpr{
+      std::get<parser::Scalar<parser::Expr>>(selectCaseStmt.statement.t).thing};
+  std::optional<TypeCategory> selectCaseStmtType;
+  int selectCaseStmtKind;
+  bool validCaseStmtType{false};
+  if (const auto *expr{GetExpr(parsedExpr)}) {
+    if (auto type{expr->GetType()}) {
+      selectCaseStmtType = type->category();
+      selectCaseStmtKind = type->kind();
```

Comment:
Why not just save the type rather than selectCaseStmtType and SelectCaseStmtKind? That lets the reader know that the category and kind are tied together. `type->category()` is clearer and shorter than `selectCaseStmtType.value()`.

Reply:
Thank you. This is a better approach. I will modify to keep the type instead of category and kind.